http/src/v0/refs/path.rs

```diff
@@ -0,0 +1,390 @@
+//! Ipfs path handling following https://github.com/ipfs/go-path/. This rests under ipfs-http for
+//! now until it's ready to be moved to `rust-ipfs`. There is a competing implementation under
+//! `libipld` which might do almost the same things, but with different dependencies. This should
+//! be moving over to `ipfs` once we have seen that this works for `api/v0/dag/get` as well.
+//!
+//! Does not allow the root to be anything else than `/ipfs/` or missing at the moment.
+
+use libipld::cid::{self, Cid};
+use libipld::Ipld;
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+use std::fmt;
+
+#[derive(Debug)]
+pub enum PathError {
+    InvalidCid(cid::Error),
+    InvalidPath,
+}
+
+impl fmt::Display for PathError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            PathError::InvalidCid(e) => write!(fmt, "{}", e),
+            PathError::InvalidPath => write!(fmt, "invalid path"),
+        }
+    }
+}
+
+impl std::error::Error for PathError {}
+
+/// Ipfs path following https://github.com/ipfs/go-path/
+#[derive(Debug)]
+pub struct IpfsPath {
+    /// Option to support moving the cid
+    root: Option<Cid>,
+    path: std::vec::IntoIter<String>,
+}
+
+impl From<Cid> for IpfsPath {
+    /// Creates a new IpfsPath from just the Cid, which is the same as parsing from a string
+    /// representation of a Cid but cannot fail.
+    fn from(root: Cid) -> IpfsPath {
+        IpfsPath {
+            root: Some(root),
+            path: Vec::new().into_iter(),
+        }
+    }
+}
+
+impl TryFrom<&str> for IpfsPath {
+    type Error = PathError;
+
+    fn try_from(path: &str) -> Result<Self, Self::Error> {
+        let mut split = path.splitn(2, "/ipfs/");
+        let first = split.next();
+        let (_root, path) = match first {
+            Some("") => {
+                /* started with /ipfs/ */
+                if let Some(x) = split.next() {
+                    // was /ipfs/x
+                    ("ipfs", x)
+                } else {
+                    // just the /ipfs/
+                    return Err(PathError::InvalidPath);
+                }
+            }
+            Some(x) => {
+                /* maybe didn't start with /ipfs/, need to check second */
+                if split.next().is_some() {
+                    // x/ipfs/_
+                    return Err(PathError::InvalidPath);
+                }
+
+                ("", x)
+            }
+            None => return Err(PathError::InvalidPath),
+        };
+
+        let mut split = path.splitn(2, '/');
+        let root = split
+            .next()
+            .expect("first value from splitn(2, _) must exist");
+
+        let path = split
+            .next()
+            .iter()
+            .flat_map(|s| s.split('/').filter(|s| !s.is_empty()).map(String::from))
```

Comment:
the is_empty check was meant to remove the trailing slash but here of course it allows for `/ipfs/cid/////a` to be equivalent to `/ipfs/cid/a` ... maybe it should be an error.

Reply:
Any guidance from the spec on this? Probably fine for now and easy to fix if there's a conformance issue.

Reply:
i think we dedupe the forward slashs, let me check

Reply:
ok js dedupes everything, go-ipfs throws for multiple slash but ignores trailing

Reply:
Thanks @hugomrdias. I propose we leave the de-duping in and change it IFF there is ever an issue with it

Reply:
I'll add this as a test case explaining the situation linking Hugos comments.